File: src/core/Grid.ts
export type Cell = { face: number; u: number; v: number };
export type Direction = 'up' | 'down' | 'left' | 'right';

import type { IShapeAdapter } from '../shapes/IShapeAdapter';

export class Grid {
  constructor(
    public size: number,
    private adapter: IShapeAdapter
  ) {}

  randomCell(excluded: Cell[]): Cell {
    const cells: Cell[] = [];
    for (let face = 0; face < this.adapter.getFaceCount(); face++) {
      for (let u = 0; u < this.size; u++) {
        for (let v = 0; v < this.size; v++) {
          if (
            !excluded.some((c) => c.face === face && c.u === u && c.v === v)
          ) {
            cells.push({ face, u, v });
          }
        }
      }
    }
    if (cells.length === 0) {
      throw new Error('Grid is full; no available cells');
    }
    return cells[Math.floor(Math.random() * cells.length)];
  }

  getNeighbor(cell: Cell, dir: Direction): Cell {
    const { face } = cell;
    let { u, v } = cell;
    switch (dir) {
      case 'up':
        v -= 1;
        break;
      case 'down':
        v += 1;
        break;
      case 'left':
        u -= 1;
        break;
      case 'right':
        u += 1;
        break;
    }
    if (u < 0 || u >= this.size || v < 0 || v >= this.size) {
      return this.adapter.wrap({ face, u, v }, dir);
    }
    return { face, u, v };
  }

}

File: python/core/grid.py
from __future__ import annotations

from dataclasses import dataclass
from typing import List

from ..shapes.ishape_adapter import Cell, IShapeAdapter


direction_type = str  # 'up' | 'down' | 'left' | 'right'


@dataclass
class Grid:
    size: int
    adapter: IShapeAdapter

    def random_cell(self, excluded: List[Cell]) -> Cell:
        cells: List[Cell] = []
        for face in range(self.adapter.get_face_count()):
            for u in range(self.size):
                for v in range(self.size):
                    if not any(
                        c.face == face and c.u == u and c.v == v for c in excluded
                    ):
                        cells.append(Cell(face, u, v))
        if not cells:
            raise ValueError("Grid is full; no available cells")
        from random import choice

        return choice(cells)

    def get_neighbor(self, cell: Cell, direction: direction_type) -> Cell:
        face = cell.face
        u = cell.u
        v = cell.v
        if direction == "up":
            v -= 1
        elif direction == "down":
            v += 1
        elif direction == "left":
            u -= 1
        elif direction == "right":
            u += 1
        if u < 0 or u >= self.size or v < 0 or v >= self.size:
            return self.adapter.wrap(Cell(face, u, v), direction)
        return Cell(face, u, v)
